lib/action_subscriber/message_retry.rb

```diff
@@ -0,0 +1,38 @@
+module ActionSubscriber
+  module MessageRetry
+    def self.redeliver_message_with_backoff(env)
+      ttl = calculate_ttl(env)
+      return unless ttl
+      with_exchange(env, ttl) do |exchange|
+        exchange.publish(env.encoded_payload,
+          :content_type => env.content_type,
+          :routing_key => env.routing_key,
+          :expiration => ttl
+        )
+      end
+    end
+
+    # Private Implementation
+    def self.calculate_ttl(env)
+      death = env.headers.fetch("x-death", [{"original-expiration" => "20"}]).first
+      next_ttl = death["original-expiration"].to_s.to_i * 5
+      return false if next_ttl > 86_400_000
+      next_ttl
+    end
+
+    def self.with_exchange(env, ttl)
+      retry_name = "#{env.exchange}_retry_#{ttl}"
```

Comment:
one clarification here, the reason we have to generate an exchange and queue pair for each separate timeout period and exchange name is because the queue can only specify one dead letter exchange (where the message will published when it expires) and messages only get expired when they are at the head of a queue. So if there were a long TTL message ahead of a shorter TTL message, the shorter one would not expire until the longer one expired.

So keeping a unique queue for each period is necessary and keeping a unique queue for each exchange is necessary. I don't think that having extra exchanges and queues will be a significant overhead, but we should probably test it to make sure that is the case.